session/auth/authprovider/authprovider.go

```diff
@@ -62,6 +63,16 @@ func (ap *authProvider) Register(server *grpc.Server) {
 }
 
 func (ap *authProvider) FetchToken(ctx context.Context, req *auth.FetchTokenRequest) (rr *auth.FetchTokenResponse, err error) {
+	ac, err := ap.getAuthConfig(req.Host)
```

Comment:
Calling `authConfig` isn't really free because this should be the call that calls credentials helpers. Eg. for `gcloud` credential helper iirc this call takes hundreds of milliseconds. So we should avoid the cases where this gets called multiple times (atm seems to be called again on line 76).

Reply:
Thanks @tonistiigi I'll refactor

Reply:
I've introduced an `authConfigCache` field to avoid the redundant (possibly expensive) calls to `GetAuthConfig`.

Reply:
Isn't it simpler if `credentials()` just takes `*types.AuthConfig` as parameter or do you see that it already gets called too much?

Reply:
I had thought about doing that, and it certainly would be simpler. I can't say under what circumstances my change would be more performant.

However, what might be a real improvement here would be to call `GetAuthConfigs()` to retrieve auth sections for all repos and set that as `authConfigCache`. That way, even if a solve is interacting with multiple registries, the retrieval of auth configuration only happens once. This is from my understanding of how the solver sessions work, however, which is very naive at the moment.

What do you think? Shall I expand the change to cache all auth config sections at simplify the single host case to use a parameter?

Reply:
Looking again, I do think that the cache map is better than adding a parameter to `credentials()`. The latter has three different callers and FWICT each caller would have to independently call `GetAuthConfig` before passing the result to `credentials()`.

Reply:
> However, what might be a real improvement here would be to call `GetAuthConfigs()` to retrieve auth sections for all repos and set that as `authConfigCache`. That way, even if a solve is interacting with multiple registries, the retrieval of auth configuration only happens once. This is from my understanding of how the solver sessions work, however, which is very naive at the moment.

I'm also second guessing this approach. I see now that `GetAuthConfigs()` [does not use the credential store at all](https://github.com/docker/cli/blob/v20.10.16/cli/config/configfile/file.go#L150). Strange.